lib/evse_security/crypto/openssl/openssl_supplier.cpp

```diff
@@ -158,176 +129,184 @@ static bool export_key_internal(const KeyGenerationInfo& key_info, const EVP_PKE
     return true;
 }
 
-static bool generate_key_internal_tpm(const KeyGenerationInfo& key_info, EVP_PKEY_ptr& out_key) {
-#if EVSE_OPENSSL_VER_3
-    // Acquire TPM context
-    TPMScopedProvider tpm_provider;
-
-    // Generate TPM key. For reference see:
-    // https://github.com/Infineon/optiga-tpm-cheatsheet/blob/master/openssl3-lib-general-examples/examples.c#L104
-
-    OSSL_PARAM params[2];
-    EVP_PKEY_CTX_ptr ctx;
-
-    CryptoKeyType type = key_info.key_type;
-    if (type == CryptoKeyType::RSA_TPM20) {
-        unsigned int bits = 2048;
-        params[0] = OSSL_PARAM_construct_uint("bits", &bits);
-        ctx = EVP_PKEY_CTX_ptr(EVP_PKEY_CTX_new_from_name(NULL, "RSA", "provider=tpm2"));
-    } else if (type == CryptoKeyType::EC_prime256v1 || type == CryptoKeyType::EC_secp384r1) {
-        char* group_256 = "P-256";
-        char* group_384 = "P-384";
+constexpr const char* kt_rsa = "RSA";
+constexpr const char* kt_ec = "EC";
+
+static bool s_generate_key(const KeyGenerationInfo& key_info, KeyHandle_ptr& out_key, EVP_PKEY_CTX_ptr& ctx) {
+
+    unsigned int bits = 0;
+    char group_256[] = "P-256";
+    char group_384[] = "P-384";
+    char* group = nullptr;
+    std::size_t group_sz = 0;
+    int nid = NID_undef;
+
+    bool bResult = true;
+    bool bEC = true;
+
+    // note when using tpm2 some key_types may not be supported.
```

Comment:
On my laptop except 2048 bits, no keys are supported. Instead of relying on trial & error to generate the key, could we only use the keys for TPM that are covered by the TPM spec? For example 2048 keys are always supported. And instead of generating a 'normal' key, maybe fail in a verbose manner?

Reply:
Related to: https://github.com/EVerest/libevse-security/pull/42#discussion_r1489567167

Reply:
Where a key type isn't supported there is a log raised (and additional information is in the journald log). The CryptoKeyType variable could be a yaml or OCPP configuration variable. At the moment it is fixed and set in evse_security.cpp so any issues should be visible during testing on a device.

Reply:
Ok, so it means in that case that we need to test on a real device. Will have to keep that in mind.